### PR TITLE
Variable FPS - Render the game as fast as your GPU can

### PR DESCRIPTION
**Pull request by quyykk**
**Feature:** Decouples the draw loop from the update loop. I didn't find an issue, but these two are loosely related: #4818 #5300

## Disclaimer

<details>
   <summary>(please read)</summary>

   This PR is a work-in-progress. I still have to fix the known problems below and while the code is cleanish, it can definitely be improved. But I think it is useful for people to try it out already and for me to get feedback.

   I'll mark this PR as "ready to review" when I'm happy with the architecture/code changes and the problems below are fixed, so feel free to wait until that happens to review.
</details>

## Feature Details

Currently, the game runs at 60fps exactly no matter how fast your CPU and GPU are. This PR changes this behavior and lets you run the game at any frame rate you want, which is either unlimited or Vsync locked at whatever your monitor supports.

**This does not change the behavior of the physics! The game still uses 60 frames per second to perform calculations.**

This PR just lets the update/physics loop and the draw loop run independently from each other. So for example, if you run the game at 120fps, then there will be 2 draw calls for each update call. If that were the only change then you'd see the game stutter a bit for odd fps. To prevent this, frames are smoothed together (interpolated).

Smoothing together frames makes the game run incredibly smooth. I'm on a laptop with 60Hz, so you'd expect me not to see any difference if I run the game at 60fps, but I do. Currently the game waits (sleeps) if your CPU and GPU is too fast, which can results in stutter because of timing imprecisions; this doesn't happen anymore.

Since the update loop is independent of the draw loop, it is possible to implement smooth slow mo without dropping frames. So I have done exactly that. Pressing the fast forward key cycles through normal -> fast forward -> slowmo.

I have also changed the "Show CPU / GPU load" option to display your FPS instead of load, so make sure to enable it.

Even if the integration tests fail, you can still use ESL2 to download this PR and try it out. :)

Any and all feedback is appreciated! Questions too if you have any. Thanks!

## What still needs to be done

- [ ] Some asteroids flicker when completely zoomed out.

- [x] Lag and input lag when frame rate drops.
    When the CPU can't keep up with the physics, the game starts to lag and input (keyboard/mouse) will get lost. The solution to this would probably involve slowing down the game (just like it is currently done without this PR).

- [x] The integration tests fail sometimes.
    This is related to the above bug. 

- [x] Go through the code to find any implicit 60fps time counting that I have missed.
- [x] The background haze probably needs to be interpolated as well.

## Technical Details

<details>
<summary>If you're interested in the technical details</summary>

The main loop instead of waiting runs an update frame when the accumulated frame time is 1000/60ms. It also plays catch up if the CPU can't keep up. The main loop is basically copied over from the famous [Fix your Timestep](https://gafferongames.com/post/fix_your_timestep/) article. Even though that article is very old, it is still the cutting edge of how to implement game loops correctly.

`RenderState` is a new class that represents all of the objects' state that is relevant for drawing (position, transformations, ...). It is filled in the `Step` functions. It stores the information as hash maps because we need to know an object's old position in the previous frame.

All of the `Draw` calls now use an interpolated frame instead of the objects' positions directly.

You might notice that I added a `dt` parameter to a `Panel`'s `Draw` function. This is needed because some things need to be updated using delta times, i.e. independent of the game frame rate (60 fps). For example, the rush mission marker that blinks on the map. If you go into fast forward or slowmo mode, then it shouldn't speed up or slow down that marker respectively. The hover timers are another example.

`BatchDrawList` (due to its nature) doesn't have a way to identify an object in the previous frame. So for those, I implemented an `Id` class that basically identifies an object.

`FrameTimer` isn't used anymore so I deleted the whole class. I still don't really understand what the CPU/GPU load means and a FPS counter (at least for this PR) seemed more useful. Although I can always readd back.
</details>

## Testing Done

I have tested this PR by doing various things like zooming in and out quickly, speeding past planets and just in general played the game for a bit. All in various combinations of slowmo, normal speed and fast forward.

## Performance Impact

This feature unfortunately doesn't come for free. Interpolation involves 2 floating point multiplications and 2 floating point additions for every object on the screen as well as a hash map lookup for every object (sometimes even 2). It also requires 2 render states to be kept in memory, which depending on the thing being stored can be as little as a single `Point` or as large as batch data.

However, I don't think that this is a big hit on performance even for lots of objects on the screen. But I didn't measure anything yet.

### CODE REVIEW

**Review by Amazinite — COMMENTED** [review]
By no means a thorough review; I've just skimmed through the changed and commented on a few things that jumped out at me.

You mention that the `dt` that is now used by the panels stands for delta time. I'd either change the name to deltaTime so that the meaning doesn't need to be assumed (as not everyone will come back to the implementation details section of this PR description to see that that's what it means) or put a comment above Panel::Draw that describes what's being passed in.

The options of "either vsync locked or unlimited" seems rather limited. Would it be possible to introduce a preference that lets you cycle through various FPS locks? e.g. default to 60 and let you cycle through 90, 120, 144 (for 144hz monitors), unlimited, 30, and back to 60. It might be that I want a higher fps but I don't want to run my GPU as fast as it will go. (The desktop I currently have access to can run Minecraft at >200 FPS with reasonable graphics settings, but I cap it at lower than that because going so high is just overkill for the monitor I have. I imagine Endless Sky would run at much more than 200 FPS if I ran it at an uncapped frame rate on that computer, but I might still want something more than 60 FPS, just not as high as the GPU can manage.)

File: source/AsteroidField.cpp [hunk 1]
```diff
@@ -201,21 +203,34 @@ AsteroidField::Asteroid::Asteroid(const Sprite *sprite, double energy)
...
+bool AsteroidField::Asteroid::Step()
 {
 	angle += spin;
 	position += velocity;
 	
 	// Keep the position within the wrap square.
 	if(position.X() < 0.)
+	{
 		position = Point(position.X() + WRAP, position.Y());
+		return true;
```
[the remark this answers is not in the document]
Comment (quyykk) [reply source/AsteroidField.cpp:215]: OMG I didn't notice. THANK YOU.

<sub>This makes the asteroids flicker more... But I think I know what the issue is now.</sub>

**Comment by quyykk** [comment]
> Would it be possible to introduce a preference that lets you cycle through various FPS locks? e.g. default to 60 and let you cycle through 90, 120, 144 (for 144hz monitors), unlimited, 30, and back to 60.

Sure, but just like slow motion I'd rather do this in a separate PR.

> I imagine Endless Sky would run at much more than 200 FPS if I ran it at an uncapped frame rate on that computer

Definitely. I have an integrated graphics card from 2019 (UHD Graphics 630) and I get 400 fps.

**Comment by Amazinite** [comment]
> I still don't really understand what the CPU/GPU load means

To my understanding, the loads were the percentage of a frame that it took to calculate/draw that frame. So a CPU load of 50% means that it took half a frame's worth of time to calculate the frame. Looked at from the other direction, that means that 50% of the time between frames was spent calculating nothing, so a consistent 50% CPU load could technically have calculated at 120 FPS. 
Loads over 100% mean that it's taking longer to calculate or draw the frame than the frame lasts, meaning that 60 FPS isn't being achieved.

> Although I can always readd back.

Given that the physics still runs at 60 FPS, I honestly would like to at least keep the CPU load to get a sense of how fast the physics frames are running. Either that, or provide separate FPS values for the rendering/GPU (what you've done already) and the physics/CPU (should ideally always be at 60 FPS but may drop during big battles). I believe Roblox has something similar in one of the debug menus, with sections telling you how fast your frame rate is but also how fast the game physics are running.
image

**Comment by Terin** [comment]
FPS locks for common refresh rates would be appreciated, yeah, since my janky Arch Linux laptop setup/video drivers don't allow for the current vsync option to work. Makes the fans spin up a whole bunch!

Really looking forward to this being implemented in the game, though. Having the option for slow motion would be great for accessibility, and it looks incredibly smooth on my 165Hz monitor now. Would love to no longer have to hack my main.cpp to remove the wait between frames.

File: source/Id.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#ifndef ID_H_
+#define ID_H_
+
+
+
+unsigned NewID();
```
[the remark this answers is not in the document]
Comment (quyykk) [reply source/Id.h:18]: Should be a private static function you're right.

File: source/RenderState.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+#include <algorithm>
+#include <cassert>
+#include <sstream>
+
+using namespace std;
+
+
+
+RenderState RenderState::states[2];
+RenderState RenderState::interpolated;
```
[the remark this answers is not in the document]
Comment (quyykk) [reply source/RenderState.cpp:29]: Static data variables need to have a definition.

[the remark this answers is not in the document]
Comment (Amazinite) [reply source/RenderState.cpp:29]: I believe such variables are typically defined in the header file.

[the remark this answers is not in the document]
Comment (Amazinite) [reply source/RenderState.cpp:29]: Checks out with out Command does it. Static variables are uncommon in the project, so I didn't recognize it.

[the remark this answers is not in the document]
Comment (quyykk) [reply source/RenderState.cpp:29]: I don't like it either tbh. In C++17 I'd just add an `inline` but we're stuck on C++11 :/

**Comment by quyykk** [comment]
Thanks for the explanation! I have added the CPU load percentage back. The GPU fps number is basically the total fps, so it also includes the physics loops. I think it's fine?

**Comment by Terin** [comment]
Can MapPanel::Step() be interpolated with deltaTime, as well? Having the screen recenter is very noticeable in slow-motion.

Being able to standardize these various UI functions, which ran differently in normal and fast-motion, is another really great part of this PR.

**Comment by Terin** [comment]
Thanks! Looks like the vector doesn't line up right unless it's exactly 60 FPS, though. Above 60 will overshoot, and below will undershoot.

https://user-images.githubusercontent.com/4471575/139894069-c3f78c53-0246-4c47-8f21-0368ede01ced.mp4

**Comment by Ferociousfeind** [comment]
Well that's just because it's coded terribly. It should take the current and target locations, and interpolate between them based on the time it is now and the time it should take to interpolate, AKA repeatedly setting the position, rather than repeatedly changing the position based on similar inputs

Not only does a non-60 FPS mess with that, but some floating point rounding errors (or something similar) cause it to center improperly when jumping large distances from one point to the other. Get Omnis, then click on a mission that is around Omnis, then click it again. You'll see it'll move on the second click too, because it doesn't properly center over such a distance. This is unrelated to this PR though

**Comment by quyykk** [comment]
Well originally I didn't want to add too many things, since this PR has enough changes already. But I guess I could just leave it in yeah.

**Comment by Terin** [comment]
Any other known issues with this PR? Asteroid flickering appears to be fixed on my end, and the CI/CD error is just adding Id/RenderState to EndlessSkyLib.cbp. Couple of minor merge conflicts, and it seems good to me.

**Comment by quyykk** [comment]
> Asteroid flickering appears to be fixed on my end

It shouldn't lol. But anyways, I'm not too happy with the architecture and so I'll need to find some time to rewrite parts of this PR.

File: source/Id.cpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
+/* Id.cpp
+Copyright (c) 2021 by quyykk
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
Comment (tibetiroka) [inline source/Id.cpp:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```
The standard copyright header was changed recently.

File: source/Id.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
+/* Id.h
+Copyright (c) 2021 by quyykk
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
Comment (tibetiroka) [inline source/Id.h:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

File: source/RenderState.h [hunk 6]
```diff
@@ -0,0 +1,94 @@
+/* RenderState.h
+Copyright (c) 2021 by quyykk
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
Comment (tibetiroka) [inline source/RenderState.h:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

**Review by tibetiroka — CHANGES_REQUESTED** [review]

**Comment by quyykk** [comment]
I'll split this PR into pieces and PR those individually.

**Pull request closed without merging**